controllers/clustercache/cluster_accessor.go

```diff
@@ -0,0 +1,492 @@
+// Controller-runtime guarantees that the ClusterCache reconciler will never reconcile the same Cluster concurrently.
+// Thus, we can rely on this method not being called concurrently for the same Cluster / clusterAccessor.
+// This means we don't have to hold the lock when we create the client, cache, etc.
+// It is extremely important that we don't lock during connection creation, so that other methods like GetClient (that
+// use a read lock) can return immediately even when we currently create a client. This is necessary to avoid blocking
+// other reconcilers (e.g. the Cluster or Machine reconciler).
+func (ca *clusterAccessor) Connect(ctx context.Context) (retErr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if ca.Connected(ctx) {
+		log.V(6).Info("Skipping connect, already connected")
+		return nil
+	}
+
+	log.Info("Connecting")
+
+	// Creating clients, cache etc. is intentionally done without a lock to avoid blocking other reconcilers.
+	connection, err := ca.createConnection(ctx)
+
+	ca.lock(ctx)
+	defer ca.unlock(ctx)
+
+	defer func() {
+		if retErr != nil {
+			log.Error(retErr, "Connect failed")
+			ca.lockedState.lastConnectionCreationErrorTimestamp = time.Now()
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	log.Info("Connected")
+
+	// Only generate the clientCertificatePrivateKey once as there is no need to regenerate it after disconnect/connect.
+	// Note: This has to be done before setting connection, because otherwise this code wouldn't be re-entrant if the
+	// private key generation fails because we check Connected above.
+	if ca.lockedState.clientCertificatePrivateKey == nil {
+		log.V(6).Info("Generating client certificate private key")
+		clientCertificatePrivateKey, err := certs.NewPrivateKey()
+		if err != nil {
+			return errors.Wrapf(err, "error creating client certificate private key")
+		}
+		ca.lockedState.clientCertificatePrivateKey = clientCertificatePrivateKey
+	}
+
+	now := time.Now()
+	ca.lockedState.healthChecking = clusterAccessorLockedHealthCheckingState{
+		// A client was just created successfully, so let's set the last probe times.
+		lastProbeTimestamp:        now,
+		lastProbeSuccessTimestamp: now,
+		consecutiveFailures:       0,
+	}
+	ca.lockedState.connection = &clusterAccessorLockedConnectionState{
+		restConfig:   connection.RESTConfig,
+		restClient:   connection.RESTClient,
+		cachedClient: connection.CachedClient,
+		cache:        connection.Cache,
+		watches:      sets.Set[string]{},
+	}
+
+	return nil
+}
+
+// Disconnect disconnects a connection to the workload cluster.
+func (ca *clusterAccessor) Disconnect(ctx context.Context) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if !ca.Connected(ctx) {
+		log.V(6).Info("Skipping disconnect, already disconnected")
+		return
+	}
+
+	ca.lock(ctx)
+	defer ca.unlock(ctx)
+
+	log.Info("Disconnecting")
+
+	// Stopping the cache is non-blocking, so it's okay to do it while holding the lock.
```

Comment:
Question: And when creating a new connection we don't have to wait for this to finish right?

Reply:
I think so, because it's an entirely different cache then. I also don't know if there is a way to wait

Reply:
Some additional context. cache.stop() just cancels the context that we passed in on cache start.

This context was also used to create informers (talking about client-go informers here).

I observed via debugger that the informers eventually terminate because of the cancelled context.

I'm not aware of anything on the controller-runtime level that is expoed that we could use to confirm that everything shut down